### PR TITLE
build packages via github actions

### PR DESCRIPTION
**Pull request by bazsi**
This PR is an attempt to build tarball/rpm/deb packages as part of our CI run, this time via github actions.

Couple of notes:
- this will generate a tarball and then compile the binaries solely from the generated tarball, this way double-checking that our release tarballs are self contained in this regard. (there are fixes in this branch as they were not)
- I'd have loved to not even checking out the source from git for the binary step, but in order to do a cache-image-<OS> we need git history (or we would _always_ build docker images, making the validation a lot longer). I've tried to do this in a separate job, but seems a bit complex to me.
- so now the package steps will clone the git repo, check if we need to rebuild docker images, rm -rf all the source and continue with the tarball (and the docker image).
- one way to improve that is to create the docker image in a separate job (w/ access to the git repo) and propagate the image with a github mechanism (e.g. artifacts, packages or cache). This didn't seem simple, so I haven't done that. Once we have pushed docker images based on this case to dockerhub, the build times should improve (as at this point it correctly detects that a rebuild is needed and is one is done).
- the workflow takes ~6minutes, 2m for the tarball, and 4m in parallel for (centos-7, ubuntu-focal) packages.
- the artifacts (tarball, rpm/deb packages) are saved to github and is saved for 90 days at least if we fit into the 5GB limit of the free tier.

If we like what we see here, we could possibly retire the dbld build stuff on ci.syslog-ng.com.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by alltilla** [comment]
@bazsi 
Thanks for this PR, personally, I really like this direction.

About avoiding checking out the source:
We might be able to only check out the source, if we need to rebuild the docker image.
 1. We can create a job which only checks if we need to rebuild, but does not rebuild automatically. We can set its conclusion as a job `output`: https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idoutputs
 2. We can use this output in the `packages` job, and only checkout -> build -> rm, if we really need to do so.

I've read, that using the github methods (artifacts or packages/cache) is usually slower than pushing to and pulling from dockerhub.

Alternatively, caching via dockerhub could work, I think. We could have a separate repo (like `syslog-ng-ci-cache`) and use commit-based tags, which we would clean up frequently. The `output` logic can help here also, to determine if we need to use the latest official tag or our custom built `ci-cache` image tag. This method would completely get rid of checking out the source in the `package` job.

Anyways, I think the current solution is good enough, but if any of these ideas is interesting to you, I am happy to help.
I think this PR does not block the release, so no need to hurry.

**Review by alltilla — APPROVED** [review]

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by gaborznagy** [comment]
@kira-syslogng build packages please;

**Comment by kira-syslogng** [comment]
Package Build FAILURE

**Comment by gaborznagy** [comment]
The kira build-packages job failed due to some environmental issues. Might be due to a recent update.

**Review by gaborznagy — APPROVED** [review]

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: .github/workflows/draft-release.yml
Comment (bazsi) [reply .github/workflows/draft-release.yml:59]: I've moved this to a separate PR #3501 , to keep this focusing only the building of binary packages. I also push another fix that would drop the backup copy of the tarball (e.g. .orig we were not using that in anyway).

Comment (gaborznagy) [reply .github/workflows/draft-release.yml:59]: Can you rebase this branch, please? I see that patches of #3501 are still on this branch.
After that, I think we can merge this PR. :+1:

**Review by gaborznagy — APPROVED** [review]

**Comment by gaborznagy** [comment]
@kira-syslogng build packages please;

**Comment by gaborznagy** [comment]
AFAIK kira build packages is working again, I've just started a build (maybe the last one ever?) to see if every platform is supported.
From now on, we should increase the coverage of the github actions build packages jobs.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Package Build SUCCESS

Packages are available at https://ci.syslog-ng.com/jenkins/job/kira-dbld-pull-requests/734/artifact/

**Review by alltilla — APPROVED** [review]

**Comment by alltilla** [comment]
We could add this to nightly too, but that can go in a different PR. This one has a lot of benefits even without running nightly.

**Review by szemere — APPROVED** [review]

**Pull request merged**
